File: tests/contrato.py
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "src"))

from estacao import (  # noqa: E402
    FalhaCalibracao, FalhaLeitura, FonteConstante, FonteNivel, PainelFixo,
    SensorNivel, Sessao, adquirir, executar_ciclo,
)


class FonteQuebrada:
    def valor(self):
        raise RuntimeError("defeito inesperado")

    def unidade(self):
        return "%"


def conferir(condicao, mensagem):
    if not condicao:
        raise AssertionError(mensagem)


def main():
    sensor = SensorNivel(10)
    painel = PainelFixo(sensor)
    real = FonteNivel(sensor)
    simulada = FonteConstante()
    sessao = Sessao()
    conferir(painel.leitura() == 10, "GUIADO: painel deve consultar o sensor associado (10)")
    sensor.atualizar(20)
    conferir(painel.leitura() == 20, "GUIADO: painel deve refletir a atualizacao (20)")
    conferir(real.valor() == 20 and simulada.valor() == 42.5,
             "CONTRATO: fontes diferentes devem responder pela mesma interface")
    conferir(executar_ciclo(real, True, True, sessao) == (True, 20, "")
             and sessao.abertas == 0, "CONTRATO: leitura valida e sessao liberada")
    conferir(executar_ciclo(real, False, True, sessao) == (False, 0, "indisponivel")
             and sessao.abertas == 0,
             "CONTRATO: indisponibilidade classificada e sessao liberada")
    try:
        adquirir(real, False, False, sessao)
        conferir(False, "GUIADO: indisponibilidade deve ter prioridade")
    except FalhaCalibracao:
        conferir(False, "GUIADO: indisponibilidade tem prioridade sobre calibracao")
    except FalhaLeitura:
        pass
    conferir(sessao.abertas == 0, "CONTRATO: sessao liberada apos propagacao")
    try:
        adquirir(real, True, False, sessao)
        conferir(False, "GUIADO: falta de calibracao deve lancar FalhaCalibracao")
    except FalhaCalibracao:
        pass
    conferir(sessao.abertas == 0, "EXTENSAO: sessao liberada apos FalhaCalibracao")
    conferir(executar_ciclo(real, True, False, sessao) == (False, 0, "calibracao")
             and sessao.abertas == 0,
             "EXTENSAO: capture FalhaCalibracao antes de FalhaLeitura e informe calibracao")
    try:
        executar_ciclo(FonteQuebrada(), True, True, sessao)
        conferir(False, "CONTRATO: defeito inesperado nao pode ser convertido em ausencia")
    except RuntimeError:
        pass
    conferir(sessao.abertas == 0, "CONTRATO: sessao liberada apos defeito inesperado")
    conferir(executar_ciclo(real, True, True, sessao) == (True, 20, "")
             and sessao.abertas == 0, "CONTRATO: proximo ciclo continua apos falha")
    print("OK pratica integrada A (Python)")


if __name__ == "__main__":
    main()

File: tests/contrato.cpp
#include "estacao.hpp"
#include <iostream>
#include <stdexcept>

void conferir(bool condicao, const char* mensagem) {
    if (!condicao) throw std::runtime_error(mensagem);
}

class FonteQuebrada final : public IFonteLeitura {
public:
    double valor() const override { throw std::logic_error("defeito inesperado"); }
    const char* unidade() const override { return "%"; }
};

int main() {
    try {
        SensorNivel sensor{10};
        PainelFixo painel{sensor};
        FonteNivel real{sensor};
        FonteConstante simulada;
        int abertas = 0;
        conferir(painel.leitura() == 10, "GUIADO: painel deve consultar o sensor associado (10)");
        sensor.atualizar(20);
        conferir(painel.leitura() == 20, "GUIADO: painel deve refletir a atualizacao (20)");
        conferir(real.valor() == 20 && simulada.valor() == 42.5,
                 "CONTRATO: fontes diferentes devem responder pela mesma interface");
        conferir(executarCiclo(real, true, true, abertas).valor == 20 && abertas == 0,
                 "CONTRATO: leitura valida e sessao liberada");
        conferir(!executarCiclo(real, false, true, abertas).sucesso && abertas == 0,
                 "CONTRATO: indisponibilidade tratada e sessao liberada");
        try {
            adquirir(real, false, false, abertas);
            conferir(false, "EXTENSAO: indisponibilidade deve ter prioridade");
        } catch (const FalhaCalibracao&) {
            conferir(false, "EXTENSAO: indisponibilidade tem prioridade sobre calibracao");
        } catch (const FalhaLeitura&) {}
        conferir(abertas == 0, "CONTRATO: sessao liberada apos propagacao");
        try {
            adquirir(real, true, false, abertas);
            conferir(false, "EXTENSAO: falta de calibracao deve lancar FalhaCalibracao");
        } catch (const FalhaCalibracao&) {}
        conferir(abertas == 0, "EXTENSAO: sessao liberada apos FalhaCalibracao");
        conferir(!executarCiclo(real, true, false, abertas).sucesso && abertas == 0,
                 "EXTENSAO: cliente recupera falha de calibracao");
        FonteQuebrada quebrada;
        try {
            executarCiclo(quebrada, true, true, abertas);
            conferir(false, "CONTRATO: defeito inesperado nao pode ser convertido em ausencia");
        } catch (const std::logic_error&) {}
        conferir(abertas == 0, "CONTRATO: sessao liberada apos defeito inesperado");
        conferir(executarCiclo(real, true, true, abertas).valor == 20 && abertas == 0,
                 "CONTRATO: proximo ciclo continua apos falha");
        std::cout << "OK pratica integrada A (C++)\n";
    } catch (const std::exception& erro) {
        std::cerr << erro.what() << '\n';
        return 1;
    }
}
